File: BFS/BFS.cpp
#include <iostream>
#include <unordered_set>
#include <queue>

using namespace std;

class AdjSet
{
public:
    AdjSet(int V)
    {
        this->numVertices = V;
        adj = new unordered_set<int>[V];
        visited = new bool[V];
    }
    ~AdjSet()
    {
        delete[] visited;
        delete[] adj;
    }
    void addEdge(int, int);
    void printGraph();
    void BFS(int start);

private:
    int numVertices;
    bool *visited;
    unordered_set<int> *adj;
};

void AdjSet::addEdge(int src, int dest)
{
    adj[src].insert(dest);
}

void AdjSet::printGraph()
{
    int v = numVertices;
    for (int i = 0; i < v; i++)
    {
        cout << "\n Vertex " << i << ":";
        for (auto x : adj[i])
            cout << "-> " << x;
        printf("\n");
    }
}

void AdjSet::BFS(int start)
{
    visited[start] = true;
    queue<int> q;
    q.push(start);
    printf("\n DFS: %d-> ", start);
    while (!q.empty())
    {
        start = q.front();
        q.pop();
        if (!visited[start])
        {
            cout << start << "-> ";
            visited[start] = true;
        }

        for (auto i : adj[start])
        {
            if (!visited[i])
                q.push(i);
        }
    }
}

int main()
{
    AdjSet adj(5);
    adj.addEdge(0, 1);
    adj.addEdge(0, 2);
    adj.addEdge(0, 3);
    adj.addEdge(1, 2);
    adj.addEdge(3, 4);
    adj.printGraph();
    adj.BFS(0);
    return 0;
}

File: BFS/BFS.js
class Graph{
    constructor(V){
        this.V = V
        this.adj = {}
        for (let i = 0; i < V; i++){
            this.adj[i] = []
        }
    }

    addEdge(src, dest){
        this.adj[src].push(dest)
    }

    printGraph(){
        for (let i = 0; i < this.V; i++){
            let output = `\n Vertex ${i}:`
            for(let x = 0; x < this.adj[i].length; x++){
                output += `-> ${this.adj[i][x]}`
            }
            console.log(output)
        }
    }

    BFS(start){
        let visited = []
        for(let i = 0; i < this.V; i++)
            visited.push(false)
        let q = []
        q.push(start)
        let output = `\n BFS: `
        
        while(q.length){
            start = q.shift()

            if(!visited[start]){
                output += `${start}-> `
                visited[start] = true
            }

            for (let i = 0; i < this.adj[start].length; i++){
                if(!visited[this.adj[start][i]])
                    q.push(this.adj[start][i])
            }
        }
        console.log(output)
    }
}

let adj = new Graph(5)
adj.addEdge(0, 1)
adj.addEdge(0, 2)
adj.addEdge(0, 3)
adj.addEdge(1, 2)
adj.addEdge(3, 4)
adj.printGraph()
adj.BFS(0)